### PR TITLE
Fix IsShared bug when editRange removes multiple layers

### PR DESCRIPTION
**Pull request by CraigMacomber**
When editRange deletes enough values that it removes two layers of the tree, isShared can be set incorrectly on the new root, resulting in clones being modified when copy on write should have occurred instead.

This issue only matters when the new root is not already marked isShared, which can only occur if it's parent was not already cloned, and thus the bug can only occur when two levels of the tree are removed (because editRange always clones a shared root) AND a child of the root, which has exactly one child of its own, is the only remaining child after editRange completes. This is a pretty rare edge case, but we managed to hit it.

This PR includes a regression test as well as a fix for this case.

Also included is a version bump of typescript to include the `||=` operator used in the fix. In the current configuration typescript complied this to a backwards compatible form in javascript, so no changes to the requires ECMA script version should be involved.

### CODE REVIEW

**Review by noencke — APPROVED** [review]

**Comment by taylorsw04** [comment]
@qwertie this change is verified, but having you take a look would be nice as we need a new package published ASAP (this is causing failures in production).

File: b+tree.ts [hunk 1]
```diff
@@ -1094,9 +1094,16 @@ export default class BTree<K=any, V=any> implements ISortedMapF<K,V>, ISortedMap
       var r = root.forRange(low, high, includeHigh, true, this, initialCounter || 0, onFound);
       return typeof r === "number" ? r : r.break!;
     } finally {
-      while (root.keys.length <= 1 && !root.isLeaf)
+      let isShared: undefined | true = undefined;
```
Comment (qwertie) [inline b+tree.ts:1097]: in future, you can write this as `let isShared: undefined | true;`

**Comment by qwertie** [comment]
Okay, published.
~~~
npm notice Publishing to https://registry.npmjs.org/
+ sorted-btree@1.6.1
~~~

**Comment by CraigMacomber** [comment]
> npm bumped my package-lock.json "lockfileVersion" from 1 to 2. Hopefully that's fine to commit.

While I'm not on newer npm yet, I really should be. One more reason to upgrade is fine :) Thanks!
